File: app/src/main/java/com/fcl/plugin/mobileglues/settings/MGConfig.java
package com.fcl.plugin.mobileglues.settings;

import static com.fcl.plugin.mobileglues.MainActivity.MainActivityContext;

import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.provider.DocumentsContract;
import android.util.Log;

import com.fcl.plugin.mobileglues.MainActivity;
import com.fcl.plugin.mobileglues.utils.Constants;
import com.fcl.plugin.mobileglues.utils.FileUtils;
import com.google.gson.Gson;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import android.os.Environment;

public class MGConfig {
    private int enableANGLE;
    private int enableNoError;
    private int enableExtGL43;
    private int enableExtTimerQuery;
    private int enableExtComputeShader;
    private int maxGlslCacheSize;
    private int multidrawMode;
    private int angleDepthClearFixMode;

    public MGConfig(int enableANGLE, int enableNoError, int enableExtGL43, int enableExtTimerQuery, int enableExtComputeShader, int maxGlslCacheSize, int multidrawMode, int angleDepthClearFixMode) {
        this.enableANGLE = enableANGLE;
        this.enableNoError = enableNoError;
        this.enableExtGL43 = enableExtGL43;
        this.enableExtTimerQuery = enableExtTimerQuery;
        this.enableExtComputeShader = enableExtComputeShader;
        this.maxGlslCacheSize = maxGlslCacheSize;
        this.multidrawMode = multidrawMode;
        this.angleDepthClearFixMode = angleDepthClearFixMode;
    }

    public static MGConfig loadConfig(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            if (MainActivity.MGDirectoryUri != null) {
                try {
                    Uri configUri = DocumentsContract.buildDocumentUriUsingTree(MainActivity.MGDirectoryUri,
                            DocumentsContract.getTreeDocumentId(MainActivity.MGDirectoryUri) + "/config.json");
                    String configStr = FileUtils.readText(context, configUri);
                    return new Gson().fromJson(configStr, MGConfig.class);
                } catch (RuntimeException | IOException e) {
                    return null;
                }
            }
            return null;
        } else {
            if (!Files.exists(new File(Constants.CONFIG_FILE_PATH).toPath())) {
                return null;
            }
            try {
                String configStr = FileUtils.readText(new File(Constants.CONFIG_FILE_PATH));
                return new Gson().fromJson(configStr, MGConfig.class);
            } catch (RuntimeException | IOException e) {
                return null;
            }
        }
    }

    public int getEnableANGLE() {
        return enableANGLE;
    }

    public void setEnableANGLE(int enableANGLE) throws IOException {
        this.enableANGLE = enableANGLE;
        saveConfig();
    }

    public int getEnableNoError() {
        return enableNoError;
    }

    public void setEnableNoError(int enableNoError) throws IOException {
        this.enableNoError = enableNoError;
        saveConfig();
    }

    public int getEnableExtTimerQuery() {
        return enableExtTimerQuery;
    }

    public void setEnableExtTimerQuery(int enableExtTimerQuery) throws IOException {
        this.enableExtTimerQuery = enableExtTimerQuery;
        saveConfig();
    }

    public int getEnableExtGL43() {
        return enableExtGL43;
    }

    public void setEnableExtGL43(int enableExtGL43) throws IOException {
        this.enableExtGL43 = enableExtGL43;
        saveConfig();
    }

    public int getEnableExtComputeShader() {
        return enableExtComputeShader;
    }

    public void setEnableExtComputeShader(int enableExtComputeShader) throws IOException {
        this.enableExtComputeShader = enableExtComputeShader;
        saveConfig();
    }

    public int getMaxGlslCacheSize() {
        return maxGlslCacheSize;
    }

    public void setMaxGlslCacheSize(int maxGlslCacheSize) throws IOException {
        if (maxGlslCacheSize < -1 || maxGlslCacheSize == 0)
            return;
        if (maxGlslCacheSize == -1)
            clearCacheFile();
        this.maxGlslCacheSize = maxGlslCacheSize;
        saveConfig();
    }

    public int getMultidrawMode() {
        return multidrawMode;
    }

    public void setMultidrawMode(int multidrawMode) throws IOException {
        this.multidrawMode = multidrawMode;
        saveConfig();
    }

    public int getAngleDepthClearFixMode() {
        return angleDepthClearFixMode;
    }

    public void setAngleDepthClearFixMode(int angleDepthClearFixMode) throws IOException {
        this.angleDepthClearFixMode = angleDepthClearFixMode;
        saveConfig();
    }

    private void clearCacheFile() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            Uri cacheUri = DocumentsContract.buildDocumentUriUsingTree(MainActivity.MGDirectoryUri,
                    DocumentsContract.getTreeDocumentId(MainActivity.MGDirectoryUri) + "/glsl_cache.tmp");
            try {
                DocumentsContract.deleteDocument(MainActivityContext.getContentResolver(), cacheUri);
            } catch (IOException | RuntimeException ignored) {
            }
        } else {
            try {
                FileUtils.deleteFile(new File(Constants.GLSL_CACHE_FILE_PATH));
            } catch (IOException ignored) {
            }
        }
    }

    public void saveConfig() throws IOException {
        save(MainActivityContext);
    }

    public void saveConfig(Context context) {
        try {
            save(context);
        } catch (RuntimeException | IOException e) {
            Log.e("MG", "Failed to save the config file: " + e.getMessage());
        }
    }

    private void save(Context context) throws IOException {
        String configStr = new Gson().toJson(this);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            if (MainActivity.MGDirectoryUri == null) {
                throw new IOException("SAF directory not selected");
            }
            FileUtils.writeText(context, MainActivity.MGDirectoryUri, "config.json", configStr, "application/json");
        } else {
            FileUtils.writeText(new File(Constants.CONFIG_FILE_PATH), configStr);
        }
    }
	
	public void deleteConfig(Context context) throws IOException {
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
			if (MainActivity.MGDirectoryUri != null) {
				FileUtils.deleteFileViaSAF(context, MainActivity.MGDirectoryUri, "config.json");
			}
		} else {
			File configFile = new File(Environment.getExternalStorageDirectory(), "MG/config.json");
			if (configFile.exists()) {
				FileUtils.deleteFile(configFile);
			}
		}
	}

}

File: app/src/main/java/com/fcl/plugin/mobileglues/settings/MGConfig.kt
package com.fcl.plugin.mobileglues.settings

import android.content.Context
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import com.fcl.plugin.mobileglues.utils.Constants
import com.fcl.plugin.mobileglues.utils.FileUtils
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File
import java.io.IOException
import java.nio.file.Files
import kotlin.properties.Delegates

class MGConfig(val context: Context) {
    // 使用 Delegates.observable 委托属性
    var enableANGLE: Int by Delegates.observable(1) { _, old, new -> if (old != new) save() }
    var enableNoError: Int by Delegates.observable(0) { _, old, new -> if (old != new) save() }
    var enableExtTimerQuery: Int by Delegates.observable(1) { _, old, new -> if (old != new) save() }
    var enableExtComputeShader: Int by Delegates.observable(0) { _, old, new -> if (old != new) save() }
    var enableExtDirectStateAccess: Int by Delegates.observable(0) { _, old, new -> if (old != new) save() }
    var maxGlslCacheSize: Int by Delegates.observable(32) { _, old, new ->
        if (old != new) {
            if (new == -1) clearCacheFile()
            save()
        }
    }
    var multidrawMode: Int by Delegates.observable(0) { _, old, new -> if (old != new) save() }
    var angleDepthClearFixMode: Int by Delegates.observable(0) { _, old, new -> if (old != new) save() }
    var customGLVersion: Int by Delegates.observable(0) { _, old, new -> if (old != new) save() }
    var fsr1Setting: Int by Delegates.observable(0) { _, old, new -> if (old != new) save() }

    companion object {
        var cacheConfigPath: String? = null
        var cacheMGDir: File = File("")

        fun loadConfig(context: Context): MGConfig? {
            val configStr: String = try {
                val configFile = File(Constants.CONFIG_FILE_PATH)
                if (!Files.exists(configFile.toPath())) return null
                FileUtils.readText(configFile)
            } catch (_: Exception) {
                return null
            }

            val config = MGConfig(context)
            try {
                // 将从文件读取的配置赋值给新创建的 config 对象
                Gson().fromJson(configStr, JsonObject::class.java).apply {
                    config.enableANGLE = this.get("enableANGLE")?.asInt ?: 1
                    config.enableNoError = this.get("enableNoError")?.asInt ?: 0
                    config.enableExtTimerQuery = this.get("enableExtTimerQuery")?.asInt ?: 1
                    config.enableExtComputeShader = this.get("enableExtComputeShader")?.asInt ?: 0
                    config.enableExtDirectStateAccess =
                        this.get("enableExtDirectStateAccess")?.asInt ?: 1
                    config.maxGlslCacheSize = this.get("maxGlslCacheSize")?.asInt ?: 32
                    config.multidrawMode = this.get("multidrawMode")?.asInt ?: 0
                    config.angleDepthClearFixMode = this.get("angleDepthClearFixMode")?.asInt ?: 0
                    config.customGLVersion = this.get("customGLVersion")?.asInt ?: 0
                    config.fsr1Setting = this.get("fsr1Setting")?.asInt ?: 0
                }

                // 处理历史遗留问题
                val obj = JsonParser.parseString(configStr).asJsonObject
                if (!obj.has("enableExtTimerQuery")) config.enableExtTimerQuery = 1

                if (!obj.has("enableExtDirectStateAccess")) config.enableExtDirectStateAccess = 0
            } catch (_: Exception) {
            }

            return config
        }
    }

    fun deleteConfig() {
        try {
            val configFile = File(Constants.CONFIG_FILE_PATH)
            if (configFile.exists()) FileUtils.deleteFile(configFile)
        } catch (_: Exception) {
        }
    }

    private fun clearCacheFile() {
        (context as? LifecycleOwner)?.lifecycleScope?.launch(Dispatchers.IO) {
            try {
                FileUtils.deleteFile(File(Constants.GLSL_CACHE_FILE_PATH))
            } catch (_: Exception) {
            }
        }
    }

    @Throws(IOException::class)
    fun save() {
        val configMap = buildConfigMap()
        val configStr = Gson().toJson(configMap)
        val configFile = File(Constants.CONFIG_FILE_PATH)
        
        // 确保目录存在
        val parent = configFile.parentFile
        if (parent != null && !parent.exists()) {
            parent.mkdirs()
        }
        
        FileUtils.writeText(configFile, configStr)
    }

    fun saveToCachePath() {
        if (cacheConfigPath == null) {
            val cacheDir = context.externalCacheDir ?: context.cacheDir
            cacheMGDir = File(cacheDir, "MG")
            if (!cacheMGDir.exists()) {
                cacheMGDir.mkdirs()
            }
            cacheConfigPath = File(cacheMGDir, "config.json").absolutePath
        }

        val configMap = buildConfigMap()
        FileUtils.writeText(File(cacheConfigPath!!), Gson().toJson(configMap))
    }

    private fun buildConfigMap(): Map<String, Int> {
        return mapOf(
            "enableANGLE" to enableANGLE,
            "enableNoError" to enableNoError,
            "enableExtTimerQuery" to enableExtTimerQuery,
            "enableExtComputeShader" to enableExtComputeShader,
            "enableExtDirectStateAccess" to enableExtDirectStateAccess,
            "maxGlslCacheSize" to maxGlslCacheSize,
            "multidrawMode" to multidrawMode,
            "angleDepthClearFixMode" to angleDepthClearFixMode,
            "customGLVersion" to customGLVersion,
            "fsr1Setting" to fsr1Setting
        )
    }
}